### PR TITLE
JBIG2 Generic Encoder 

### PR DESCRIPTION
**Pull request by kucjac**
This Pull Request provide JBIG2 lossless encoder for the bi-level image compression.

---
This change is [Reviewable](https://reviewable.io/reviews/unidoc/unipdf/264)

### CODE REVIEW

File: core/encoding.go [hunk 1]
```diff
@@ -1982,180 +1981,6 @@ func (enc *CCITTFaxEncoder) EncodeBytes(data []byte) ([]byte, error) {
...
-	Globals jbig2.Globals
-	// IsChocolateData defines if the data is encoded such that
-	// binary data '1' means black and '0' white.
-	// otherwise the data is called vanilla.
-	// Naming convention taken from: 'https://en.wikipedia.org/wiki/Binary_image#Interpretation'
-	IsChocolateData bool
-}
-
-// NewJBIG2Encoder returns a new instance of JBIG2Encoder.
-func NewJBIG2Encoder() *JBIG2Encoder {
```
[the remark this answers is not in the document]
Comment (kucjac) [reply core/encoding.go]: You're definitely right. I'm reverting that change.

File: common/logging.go [hunk 2]
```diff
@@ -82,56 +86,143 @@ func (l ConsoleLogger) IsLogLevel(level LogLevel) bool {
...
+}
+
 // output writes `format`, `args` log message prefixed by the source file name, line and `prefix`
-func (l ConsoleLogger) output(f *os.File, prefix string, format string, args ...interface{}) {
+//noinspection GoUnhandledErrorResult
+func (l WriterLogger) output(f io.Writer, prefix string, format string, args ...interface{}) {
+	output(f, prefix, format, args)
+}
+
+func output(f io.Writer, prefix string, format string, args ...interface{}) {
```
Comment (adrg) [inline common/logging.go]: I think the name of this function should be changed to something a bit more suggestive. As part of a struct, it did provide some context. Maybe `logToWriter` or `logOutput` or something like that.

File: core/encoding.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (adrg) [reply core/encoding.go]: Please don't forget to bring back the `NewJBIG2Encoder` function.

File: common/logging.go [hunk 3]
```diff
@@ -82,56 +86,143 @@ func (l ConsoleLogger) IsLogLevel(level LogLevel) bool {
...
+// Trace logs 'Trace' message.
 func (l ConsoleLogger) Trace(format string, args ...interface{}) {
 	if l.LogLevel >= LogLevelTrace {
 		prefix := "[TRACE] "
 		l.output(os.Stdout, prefix, format, args...)
 	}
 }
 
+// output writes `format`, `args` log message prefixed by the source file name, line and `prefix`
+//noinspection GoUnhandledErrorResult
```
Comment (gunnsth) [inline common/logging.go]: Remove the "noinspection" comment, seems to be for some linter that we are not using.

File: common/logging.go [hunk 4]
```diff
@@ -82,56 +86,143 @@ func (l ConsoleLogger) IsLogLevel(level LogLevel) bool {
...
 
+// Warning logs 'warning' message.
 func (l ConsoleLogger) Warning(format string, args ...interface{}) {
 	if l.LogLevel >= LogLevelWarning {
 		prefix := "[WARNING] "
 		l.output(os.Stdout, prefix, format, args...)
 	}
 }
 
+// Notice logs 'Notice' message.
```
Comment (gunnsth) [inline common/logging.go]: Should be // Note logs notice messages.
No need to put 'Notice' into quotes.  Same for other comment similar.
(Error, Warning, Debug, Trace)
Sometimes variable names are put into ticks
https://github.com/unidoc/unipdf/wiki/UniPDF-Developer-Guide#12-doc-comments
but not usually function names

File: core/encoding_jbig2.go [hunk 5]
```diff
@@ -0,0 +1,651 @@
...
+	if enc.IsChocolateData {
+		// /Decode[1.0 0.0] - see note in the 'setChocolateData' method.
+		dict.Set("Decode", MakeArray(MakeFloat(1.0), MakeFloat(0.0)))
+	}
+	dict.Set("Filter", MakeName(enc.GetFilterName()))
+	return dict
+}
+
+// UpdateParams updates the parameter values of the encoder.
+func (enc *JBIG2Encoder) UpdateParams(params *PdfObjectDictionary) {
```
[the remark this answers is not in the document]
Comment (gunnsth) [reply core/encoding_jbig2.go]: check? empty or not?

File: core/encoding_jbig2.go [hunk 6]
```diff
@@ -0,0 +1,437 @@
...
+	}
+
+	if decodeParams != nil {
+		if globals := decodeParams.Get("JBIG2Globals"); globals != nil {
+			var err error
+
+			globalsStream, ok := globals.(*PdfObjectStream)
+			if !ok {
+				err = errors.Error(processName, "jbig2.Globals stream should be an Object Stream")
+				//noinspection GoNilness
```
Comment (gunnsth) [inline core/encoding_jbig2.go]: Can you remove those noinspection comments?
At the moment we just try to minimize golint output but not commenting on each single deviation.

File: go.mod [hunk 7]
```diff
@@ -12,3 +12,5 @@ require (
 	golang.org/x/text v0.3.2
 	golang.org/x/tools v0.0.0-20190606174628-0139d5756a7d // indirect
 )
+
+go 1.13
```
Comment (gunnsth) [inline go.mod]: Can you remove this?  Or  check what is the oldest version that works?
I think go 1.11 should work.

File: internal/jbig2/bitmap/bitmap_test.go [hunk 8]
```diff
@@ -251,4 +293,889 @@ func TestBitmap(t *testing.T) {
...
+			bm, err := NewWithData(20, 2, data)
+			require.NoError(t, err)
+
+			assert.False(t, bm.Zero())
+		})
+
+	})
+
+	t.Run("ConnComponents", func(t *testing.T) {
+
```
Comment (gunnsth) [inline internal/jbig2/bitmap/bitmap_test.go]: Add TODO for adding this test?

File: internal/jbig2/bitmap/combine.go [hunk 9]
```diff
@@ -76,6 +77,17 @@ func Extract(roi image.Rectangle, src *Bitmap) (*Bitmap, error) {
 	return dst, nil
 }
 
+// combineBitmap combines two bitmaps with the
```
Comment (gunnsth) [inline internal/jbig2/bitmap/combine.go]: with the ...?

File: internal/jbig2/bitmap/components.go [hunk 10]
```diff
@@ -0,0 +1,355 @@
...
+			return nil, errors.Wrap(err, processName, "bm2 -XOR-> bm3")
+		}
+		bma.AddBitmap(bm3)
+		xStart = pt.X
+		yStart = pt.Y
+	}
+	bma.Boxes = *boxa
+	return boxa, nil
+}
+func wordMaskByDilation(s *Bitmap) (mask *Bitmap, dilationSize int, err error) {
```
Comment (gunnsth) [inline internal/jbig2/bitmap/components.go]: should be empty newline above the function.
Need to run goimports ?

File: internal/jbig2/bitmap/components_test.go [hunk 11]
```diff
@@ -0,0 +1,1466 @@
...
+	})
+
+	t.Run("EmptyBitmap", func(t *testing.T) {
+		var b *Bitmap
+		_, _, err := b.GetComponents(ComponentConn, 12, 8)
+		assert.Error(t, err)
+	})
+
+	t.Run("NoEmptyFrame", func(t *testing.T) {
+		// t.Skip()
```
Comment (gunnsth) [inline internal/jbig2/bitmap/components_test.go]: Can we remove the commented code?
Typically we don't keep commented code unless for some reason, which should be indicated.

File: internal/jbig2/document/document.go [hunk 12]
```diff
@@ -0,0 +1,838 @@
...
+	if err == io.EOF {
+		return true, nil
+	} else if err != nil {
+		return false, errors.Wrap(err, processName, "")
+	}
+	return false, nil
+}
+
+// func (d *Document) uniteTemplatesWithIndexes(firstTempIndex, secondTempIndex int) error {
+// 	if len(d.Classer.Pixat) < firstTempIndex || len(d.Classer.Pixat) < secondTempIndex {
```
Comment (gunnsth) [inline internal/jbig2/document/document.go]: can remove commented code

File: internal/jbig2/document/document_test.go [hunk 13]
```diff
@@ -486,79 +499,252 @@ func TestDocument(t *testing.T) {
...
+				assert.Error(t, err)
+			})
+		})
+	})
+
+	// t.Run("Classified", func(t *testing.T) {
+	// 	t.Run("PDFMode", func(t *testing.T) {
+	// 		scales := []int{2, 4}
+	// 		for _, scale := range scales {
+	// 			t.Run(fmt.Sprintf("SinglePageX%d", scale), func(t *testing.T) {
```
Comment (gunnsth) [inline internal/jbig2/document/document_test.go]: remove commented code or add explanation if there is reason to keep

File: internal/jbig2/document/page.go [hunk 14]
```diff
@@ -0,0 +1,563 @@
...
+		}
+	}
+	return regionCount
+}
+
+// encodeSegment encodes the segment data and segment header for given 'segmentNumber'.
+// Then the function writes it's encoded data into 'w' writer.
+// func (p *Page) encodeSegment(w writer.BinaryWriter, segmentNumber int) (n int, err error) {
+//	const processName = "encodeSegment"
+//	// get the segment for given 'segmentNumber'
```
Comment (gunnsth) [inline internal/jbig2/document/page.go]: commented code, remove or add reason to keep

File: internal/jbig2/document/page.go [hunk 15]
```diff
@@ -0,0 +1,563 @@
...
+// 	n, err = seg.Encode(w)
+// 	if err != nil {
+// 		return 0, errors.Wrapf(err, processName, "page: '%d'", p.PageNumber)
+// 	}
+// 	return n, nil
+// }
+
+// firstSegmentNumber gets the number of the first segment in the page.
+// func (p *Page) firstSegmentNumber() (first uint32, err error) {
+// 	const processName = "firstSegmentNumber"
```
Comment (gunnsth) [inline internal/jbig2/document/page.go]: commented code, remove or document reason to keep

File: internal/jbig2/document/segments/symbol-dictionary.go [hunk 16]
```diff
@@ -83,44 +87,437 @@ type SymbolDictionary struct {
...
+			}
+		}
+		if err = ectx.EncodeOOB(encoder.IADW); err != nil {
+			return 0, errors.Wrap(err, processName, "IADW")
+		}
+	}
+
+	// // iterate over all symbols in the list
+	// for i := 0; i < len(s.symbolList); {
+	// 	height, err := s.bmHeight(i)
```
Comment (gunnsth) [inline internal/jbig2/document/segments/symbol-dictionary.go]: commented code: remove or document reason to keep

**Pull request merged**